cmd/kubeadm/app/phases/certs/certlist.go

```diff
@@ -0,0 +1,303 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certs
+
+import (
+	"crypto/rsa"
+	"crypto/x509"
+	"fmt"
+
+	certutil "k8s.io/client-go/util/cert"
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	"k8s.io/kubernetes/cmd/kubeadm/app/phases/certs/pkiutil"
+)
+
+type configMutatorsFunc func(*kubeadmapi.InitConfiguration, *certutil.Config) error
+
+// KubeadmCert represents a certificate that Kubeadm will create to function properly.
+type KubeadmCert struct {
```

Comment:
I would like to see something in this structure that could help in cleaning up `UsingExternalCA` and `SharedCertificateExists` e.g.
- for `UsingExternalCA` we need to know which KubeadmCert is a CA (might be func (k *KubeadmCert) IsCa returning CAName=="")
- for `SharedCertificateExists` we could rely on the same function I guess...

wdyt?

Reply:
@fabriziopandini that's for phase 2! I'm trying to keep this change small so it's easier to review, but eventually I'm going to remove all the functions in certs/ and only use the new struct